api/v1alpha1/machineconfigserver_types.go

```diff
@@ -0,0 +1,51 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func init() {
+	SchemeBuilder.Register(&MachineConfigServer{}, &MachineConfigServerList{})
+}
+
+// MachineConfigServerSpec defines the desired state of MachineConfigServer
+type MachineConfigServerSpec struct {
+
+	// Release specifies the release image to use for this MachineConfigServer
+	ReleaseImage string `json:"releaseImage"`
+}
+
+// MachineConfigServerStatus defines the observed state of MachineConfigServer
+type MachineConfigServerStatus struct {
+	// Version is the semantic version of the release used by the mcs.
+	// For a mcs a given version represents the ignition config served by
+	// the ignition endpoint referenced in the userdata secret.
+	// +kubebuilder:validation:Optional
+	Version string `json:"version,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	Host string `json:"host,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=machineconfigservers,shortName=mcs;mcss,scope=Namespaced
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Version",type="string",JSONPath=".status.version",description="Version"
+// +kubebuilder:printcolumn:name="Host",type="string",JSONPath=".status.host",description="Host"
+// MachineConfigServer is the Schema for the MachineConfigServers API
+type MachineConfigServer struct {
```

Comment:
Not entirely clear on the purpose of this new type. Is it true there's a 1:1 affinity between a MachineConfigServer and a NodePool for a specific release image value? And is it true the MachienConfigServer is purely a function of a NodePool? Then for example, could the same job be done if the ignition address a property of the NodePool status? Sorry if this is a dumb Q, trying to understand the model

Reply:
>Is it true there's a 1:1 affinity between a MachineConfigServer and a NodePool for a specific release image value? And is it true the MachineConfigServer is purely a function of a NodePool? Then for example, could the same job be done if the ignition address a property of the NodePool status? Sorry if this is a dumb Q, trying to understand the model

The functionality this PR is enabling is fine grained control over the versioning of nodePools. By having versioning for nodePools we will be able for instance orchestrate upgrades for nodePools in a cluster one by one.
To achieve this we are exposing a new building block machineConfigServer which is used at convenience here to that end but not limited to it

Reply:
Right but what I'm asking is what does the new intermediate type (MachineConfigServer) enable that isn't possible if the ignition server details were embedded in the NodePool status?

Maybe another clarifying question: when would a MachineConfigServer be deleted?

Reply:
Ah I see your point. The reason to propose having that logic in its own crd is to support the use case where machines are managed externally and not by a nodePool. In such scenarios as a cluster provider I want to create just a MCS so as a cluster admin you can bring your own machine that consumes the right ignition.

Reply:
I think we all agreed this is a useful abstraction, thanks!